mkfs/rootdir.c

```diff
@@ -362,6 +362,76 @@ static int add_symbolic_link(struct btrfs_trans_handle *trans,
 	return ret;
 }
 
+/*
+ * keep our extent size at 1MB max, this makes it easier to work
+ * inside the tiny block groups created during mkfs
+ */
+#define MAX_EXTENT_SIZE SZ_1M
+
+static int read_and_write_extent(struct btrfs_trans_handle *trans,
+				 struct btrfs_root *root,
+				 struct btrfs_inode_item *btrfs_inode,
+				 u64 objectid, int fd, u64 file_pos, char *buf,
```

Comment:
Everything from `fd`, `buf`, `size` (from st structure), and `path_name` is only here to help read the file.
The only thing changing is `file_pos` which is worth a dedicated parameter.

During my initial failed merge, I added an extra helper structure, source_descriptor, to include `fd`, `st`, `path` and `buf`. Later expands the structure to add `compressed_buf` for compression.

Such structure is more expandable than adding more and more parameters to the already long list.